### PR TITLE
update predictors to use Cog's new Pydantic API

### PR DESCRIPTION
**Pull request by zeke**
Hey @dribnet 👋🏼 

This PR updates pixray's Cog predictor classes to be compatible with [Cog's new Python API](https://github.com/replicate/cog/blob/703db05575a044da887aff2f1858e20be9d225df/docs/python.md).

Cog v0.1.0 has a new predictor API that makes use of Python's built-in type annotations to declare input and output types. The new API also has a different way of declaring inputs based on [pydantic](https://pydantic-docs.helpmanual.io/), a python package for validating Python models. Instead of using the `@cog.input` decorators, inputs are now declared inline as parameters to the `predict()` method.

There's a bunch of other useful foundational stuff in this new release of Cog that gets us closer to having a standardize type system that leans on JSON Schema and OpenAPI instead of re-inventing our own thing. For more details, see the release notes here: https://github.com/replicate/cog/releases/tag/v0.1.0

### Process

Here's the process I followed to set things up, make changes, and test:

1. created a new model https://replicate.com/zeke/pydantic-pixray
1. forked this repo, recursively cloned it, and `cog push`ed it to zeke/pydantic-pixray using "old cog" (0.0.x)
1. verified that my unchanged fork of pixray worked by running some predictions
1. upgraded my local Cog version to the latest, [`0.1.1`](https://github.com/replicate/cog/releases/tag/v0.1.1)
1. updated all the Predictors in `cogrun.py` to use `cog.BasePredictor`, `cog.Input`, `cog.Path` etc.
1. set the `image` field in cog.yaml to publish to my own copy of the model and the `predict` field to `cogrun.py:PixrayVdiff`
1. published using `cog push`
1. ran `cog predict` from a GCP image with a GPU. See https://gist.github.com/zeke/a36c059bebb751fb21b26c1d14ed1996

### Progress

I am now able to `cog build`, `cog push`, and `cog predict` the changes herein using the latest version of Cog, but still hitting a few snags:

- The `PixrayVdiff` predictor (which I think corresponds to https://replicate.com/pixray/text2image) produces output, but it's yielding the same image over and over. See https://gist.github.com/zeke/a36c059bebb751fb21b26c1d14ed1996
- Some of the existing predictors accept a `kwargs` argument, but the new version of Cog has a strict list of [allowed input types](https://github.com/replicate/cog/blob/6c6988d1ff65aef3a9fdb96f167a86972508190f/python/cog/predictor.py#L20:L20). In order to be compatible with Cog's new type checking stuff, these predictors that accept arbitrary keyword arguments will need to be expanded to explicitly list all the arguments and their types.

### Next steps

@dribnet hopefully this gives you a head start for updating pixray to work with the new version of Cog. Let me know if this all makes sense, and if you need more help getting these changes shipped.

### CODE REVIEW

File: cogrun.py [hunk 1]
```diff
@@ -15,15 +16,15 @@ def create_temporary_copy(src_path):
...
         os.environ['TORCH_HOME'] = 'models/'
 
-    # Define the input types for a prediction
-    @cog.input("settings", type=str, help="Default settings to use")
-    @cog.input("prompts", type=str, help="Text Prompts")
-    def predict(self, settings, **kwargs):
+    def predict(self, 
+        settings: str = Input(description="Default settings to use"),
+        prompts: str = Input(description="Text prompts", default=None),
+    **kwargs) -> Iterator[str]:
```
Comment (bfirsh) [inline cogrun.py:27]: Should this be `Iterator[Path]`?

Comment (zeke) [reply cogrun.py]: Yep. Fixed in https://github.com/pixray/pixray/pull/50/commits/fe66a1c94b63fdea9202dac7e814f25376bfbeae

**Comment by dribnet** [comment]
Thanks Zeke - this was a heroic mapping of my current setup and seems to be working fine locally with the latest cog.

I'm also pushing this up to replicate, but currently getting failures:
```bash
received unexpected HTTP status: 500 Internal Server Error
ⅹ exit status 1
```

I'm currently pushing to the `text2image-future` endpoint [which is here](https://replicate.com/pixray/text2image-future).
The error happens at the very end of the process and takes several hours so I'll try again and let you know if I have better luck the 2nd time.

**Comment by zeke** [comment]
Hey @dribnet glad to hear this is on the right track.

Catching up on the chat in Discord, It looks like that 500 was the result of a timeout, and you were able to circumvent it by adding some big unneeded files to your `.dockerignore`. Is that right?

**Comment by dribnet** [comment]
Yep - I seem to be OK with 5-10G of models so I think I'll be OK. Thanks again for getting this migration going for me.

**Comment by dribnet** [comment]
Thanks - working well on the [text2image-future model endpoint](https://replicate.com/pixray/text2image-future) and now scheduled to go into the next release. 👍

**Comment by zeke** [comment]
Whoop whoop!

**Pull request merged**